activerecord/test/cases/associations/has_one_through_associations_test.rb

```diff
@@ -473,4 +476,20 @@ def test_cpk_stale_target
 
     assert_predicate(book.association(:order_agreement), :stale_target?)
   end
+
+  def test_bidirectional_has_one_through_destroys_from_both_ends
```

Comment:
We should be able to simplify the issue without the need to have a `through` relation. I think there should be existing models/fixtures in the codebase to reproduce the problem:

<details>
<summary> Bug report </summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  # gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :pilots, force: true do |t|
  end

  create_table :cars, force: true do |t|
    t.integer :pilot_id
    t.integer :copilot_id
  end
end

class Pilot < ActiveRecord::Base
  has_one :car, dependent: :destroy
end

class Car < ActiveRecord::Base
  belongs_to :pilot, dependent: :destroy
  belongs_to :copilot, class_name: "Pilot", dependent: :destroy
end

class BugTest < Minitest::Test
  def test_destroy
    pilot = Pilot.create!
    copilot = Pilot.create!
    car = Car.create!(pilot: pilot, copilot: copilot)

    pilot.destroy!

    assert_raises(ActiveRecord::RecordNotFound) { car.reload }
    assert_raises(ActiveRecord::RecordNotFound) { pilot.reload }
    assert_raises(ActiveRecord::RecordNotFound, "The Copilot wasn't destroyed") { copilot.reload }
  end
end
```
</details>

Reply:
The existing models have so much stuff in them that I couldn't find any to suit my left/middle/right setup, so I added new ones.

I agree that the bug can be reproduced with two belongs_to relations pointing to the same class; we don't need two different classes on the other end of the relations.

Reply:
Out of curiosity, why do you prefer `assert_raises(ActiveRecord::RecordNotFound) { car.reload }` to `assert_predicate(car, :destroyed?)`?

Reply:
Is `Pilot` a new class?  I can find [`Car`](https://github.com/rails/rails/blob/main/activerecord/test/models/car.rb) but not `Pilot`.

Reply:
> Is Pilot a new class?

It was just an example to find a similar setup in the codebase. This two models are what you are looking for (just need to add `dependent :destroy` on both side of the association). The diff will be much smaller and we don't need to add new tables/models https://github.com/rails/rails/blob/cb47c12aa5ec907204c9e9c26fd57a86d2f0d86f/activerecord/test/models/user.rb#L16
https://github.com/rails/rails/blob/cb47c12aa5ec907204c9e9c26fd57a86d2f0d86f/activerecord/test/models/room.rb#L4-L5

> Out of curiosity, why do you prefer assert_raises(ActiveRecord::RecordNotFound) { car.reload }

No reason in this case, `assert_predicate(car, :destroyed?)` is fine too!